include/oneapi/tbb/flow_graph.h

```diff
@@ -3329,20 +3329,41 @@ class overwrite_node : public graph_node, public receiver<T>, public sender<T> {
         return false;
     }
 
+#if __TBB_PREVIEW_FLOW_GRAPH_TRY_PUT_AND_WAIT
+    bool try_get( input_type &v, message_metainfo& metainfo ) override {
+        spin_mutex::scoped_lock l( my_mutex );
+        if (my_buffer_is_valid) {
+            v = my_buffer;
+            metainfo = my_buffered_metainfo;
+
+            // Since the successor of the node will use move semantics while wrapping the matainfo
+            // that is designed to transfer the ownership of the value from single-push buffer to the task
+            // It is required to reserve one more reference here because the value keeps in the buffer
+            // and the ownership is not transferred
+            for (auto msg_waiter : metainfo.waiters()) {
+                msg_waiter->reserve(1);
+            }
+            return true;
+        }
+        return false;
+    }
+#endif
+
     //! Reserves an item
     bool try_reserve( T &v ) override {
         return try_get(v);
     }
 
 #if __TBB_PREVIEW_FLOW_GRAPH_TRY_PUT_AND_WAIT
 private:
-    // TODO: add real implementation
-    bool try_reserve(T& v, message_metainfo&) override {
-        return try_reserve(v);
-    }
-
-    bool try_get( input_type& v, message_metainfo& ) override {
-        return try_get(v);
+    bool try_reserve(T& v, message_metainfo& metainfo) override {
+        spin_mutex::scoped_lock l( my_mutex );
+        if (my_buffer_is_valid) {
+            v = my_buffer;
+            metainfo = my_buffered_metainfo;
```

Comment:
I suppose there is no move of the metainfo when reserved/consumed and so there is no need to increment ref count here or in consume. Is that right? It is handled in the downstream node.

Reply:
Yes, move semantics optimization is only applied for try_get.